src/NodeManager.php

```diff
@@ -61,41 +86,41 @@ public function generateMultiple(array $ids): array
         return array_filter($nodes, static fn ($buffer) => null !== $buffer);
     }
 
-    /**
-     * Generate the node buffer (content elements).
-     */
     private function generateBuffer(NodeModel $nodeModel): string
     {
         if (!Controller::isVisibleElement($nodeModel)) {
             return '';
         }
 
-        $buffer = '';
-        $elementsData = [];
+        $nodeElements = [];
 
         if (null !== ($elements = $nodeModel->getContentElements())) {
             /** @var ContentModel $element */
             foreach ($elements as $index => $element) {
-                $elementsData[] = $element->row();
+                // Keep the index if somebody wants to refer that inside the generated content element
                 $element->nodeElementIndex = $index;
-                $buffer .= Controller::getContentElement($element);
+
+                $nodeElements[] = new NodeElement($element->row(), Controller::getContentElement($element));
             }
         }
 
         if (!$nodeModel->wrapper) {
-            return $buffer;
+            return implode('', array_map(static fn (NodeElement $v) => $v->getRenderedHtml(), $nodeElements));
         }
 
-        $template = new FrontendTemplate($nodeModel->nodeTpl ?: 'node_default');
-        $template->setData($nodeModel->row());
-        $template->elementsData = $elementsData;
-
         $cssID = StringUtil::deserialize($nodeModel->cssID, true);
 
-        $template->class = !empty($cssID[1]) ? $cssID[1] : '';
-        $template->cssID = !empty($cssID[0]) ? $cssID[0] : '';
-        $template->buffer = $buffer;
+        if ($nodeModel->nodeTpl) {
```

Comment:
I think this should not happen in the `NodeManager`. Also, imho we don't need a a separate wrapper template. Can't we just have one template to rule it all? It can `include` based on `nodeTpl` itself? Then people could control everything in one template?